### PR TITLE
[procutil] provide options to skip collecting process fields that need extra permission

### PR DESCRIPTION
**Pull request by shang-wang**
### What does this PR do?

This PR creates more function in procutil to provide options for callers to skip collecting process fields that need extra permission.

cc: @DataDog/processes 
 
### Motivation

We plan to move the collection for fields that need extra permission into system-probe, so these newly added functions could be used by process-agent and combine the data from system-probe.

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details that should be tested during the QA.

### CODE REVIEW

File: pkg/process/checks/allprocesses_linux.go [hunk 1]
```diff
@@ -12,15 +12,15 @@ import (
 // getAllProcesses uses a probe to fetch processes using procutil library,
 // then convert them into FilledProcesses for compatibility
 func getAllProcesses(probe *procutil.Probe) (map[int32]*process.FilledProcess, error) {
-	procs, err := probe.ProcessesByPID(time.Now())
+	procs, err := probe.ProcessesByPID(time.Now(), true)
```
Comment (robertjli) [inline pkg/process/checks/allprocesses_linux.go:15]: Why not change this to `ProcessesByPIDWithPerm()`?

File: pkg/process/procutil/process_linux.go [hunk 2]
```diff
@@ -104,24 +114,39 @@ func (p *Probe) StatsForPIDs(pids []int32, now time.Time) (map[int32]*Stats, err
...
+			Nice:        statInfo.nice,          // /proc/[pid]/stat
+			CPUTime:     statInfo.cpuStat,       // /proc/[pid]/stat
+			MemInfo:     statusInfo.memInfo,     // /proc/[pid]/status
+			MemInfoEx:   memInfoEx,              // /proc/[pid]/statm
+			CtxSwitches: statusInfo.ctxSwitches, // /proc/[pid]/status
+			NumThreads:  statusInfo.numThreads,  // /proc/[pid]/status
+		}
+		if withPerm {
+			stats.OpenFdCount = p.getFDCount(pathForPID) // /proc/[pid]/fd, requires permission checks
+			stats.IOStat = ioInfo                        // /proc/[pid]/io, requires permission checks
```
Comment (robertjli) [inline pkg/process/procutil/process_linux.go:129]: Should `p.parseIO()` also be moved into this block?

Comment (shang-wang) [reply pkg/process/procutil/process_linux.go:129]: Good catch I copy pasted and messed up the order.

File: pkg/process/procutil/process_linux.go [hunk 3]
```diff
@@ -158,18 +182,24 @@ func (p *Probe) ProcessesByPID(now time.Time) (map[int32]*Process, error) {
...
+				CPUTime:     statInfo.cpuStat,       // /proc/[pid]/stat
+				MemInfo:     statusInfo.memInfo,     // /proc/[pid]/status
+				MemInfoEx:   memInfoEx,              // /proc/[pid]/statm
+				CtxSwitches: statusInfo.ctxSwitches, // /proc/[pid]/status
+				NumThreads:  statusInfo.numThreads,  // /proc/[pid]/status
 			},
 		}
+		if withPerm {
+			proc.Stats.OpenFdCount = p.getFDCount(pathForPID) // /proc/[pid]/fd, requires permission checks
+			ioInfo := p.parseIO(pathForPID)
```
Comment (mbotarro) [inline pkg/process/procutil/process_linux.go:197]: Do we need this `ioInfo` var?

**Review by mbotarro — APPROVED** [review]

File: pkg/process/checks/allprocesses_linux.go (hunk 1 above)
Comment (sunhay) [reply pkg/process/checks/allprocesses_linux.go:15]: +1 - lets keep the API as straightforward as possible.

File: pkg/process/procutil/process_linux.go [hunk 4]
```diff
@@ -89,8 +89,18 @@ func (p *Probe) Close() {
 	}
 }
 
+// StatsForPIDsWithPerm collects stats for processes that require extra permissions
+func (p *Probe) StatsForPIDsWithPerm(pids []int32, now time.Time) (map[int32]*Stats, error) {
```
Comment (sunhay) [inline pkg/process/procutil/process_linux.go:93]: I don't know if I see a use-case where someone wants to flip between these two functions. Can we just configure whether this should be collected or not in the `Probe{}` itself?

File: pkg/process/checks/allprocesses_linux.go (hunk 1 above)
Comment (sunhay) [reply pkg/process/checks/allprocesses_linux.go:15]: Actually I think we should get rid of it completely and have this configuration live in the `Probe{}` (commented below as well)

File: pkg/process/procutil/process_linux.go (hunk 4 above)
Comment (shang-wang) [reply pkg/process/procutil/process_linux.go]: Updated. That's a good choice.

**Pull request merged**